adapters/rubicon/rubicon.go

```diff
@@ -649,19 +659,74 @@ func (a *RubiconAdapter) MakeRequests(request *openrtb.BidRequest, reqInfo *adap
 				// set user.ext.tpid
 				if len(userExt.Eids) > 0 {
 					tpIds := make([]rubiconExtUserTpID, 0)
+
+					var segments []string
 					for _, eid := range userExt.Eids {
 						if eid.Source == "adserver.org" {
 							uids := eid.Uids
 							if len(uids) > 0 {
 								uid := uids[0]
-								if uid.Ext != nil && uid.Ext.RtiPartner == "TDID" {
-									tpIds = append(tpIds, rubiconExtUserTpID{Source: "tdid", UID: uid.ID})
+
+								if uid.Ext != nil {
+									var eidUidExt rubiconUserExtEidUidExt
+									if err := json.Unmarshal(uid.Ext, &eidUidExt); err != nil {
+										errs = append(errs, &errortypes.BadInput{
+											Message: err.Error(),
+										})
+										continue
+									}
+
+									if eidUidExt.RtiPartner == "TDID" {
+										tpIds = append(tpIds, rubiconExtUserTpID{Source: "tdid", UID: uid.ID})
+									}
+								}
+							}
```

Comment:
Please consider moving this `if` block and it's `else` block  into their own functions for better readability and easier unit testing. For example, in case where source is `liveintent.com`, `tpId.UID` is set to `uids[0].ID` but that's not covered in the unit test and with a unit test trying to test the entire OpenRTB request, it is hard to incorporate all the different test cases. Breaking it into smaller functions will help with better test coverage.